examples/service_explorer.py

```diff
@@ -26,35 +26,39 @@ async def run(address, debug=False):
         log.addHandler(h)
 
     async with BleakClient(address) as client:
-        x = await client.is_connected()
-        log.info("Connected: {0}".format(x))
+        is_connected = await client.is_connected()
+        log.info(f"Connected: {is_connected}")
 
         for service in client.services:
-            log.info("[Service] {0}: {1}".format(service.uuid, service.description))
+            log.info(f"[Service] {service}")
             for char in service.characteristics:
                 if "read" in char.properties:
                     try:
                         value = bytes(await client.read_gatt_char(char.uuid))
+                        log.info(
+                            f"\t[Characteristic] {char} ({','.join(char.properties)}), Value: {value}"
+                        )
                     except Exception as e:
                         value = str(e).encode()
+                        log.error(
+                            f"\t[Characteristic] {char} ({','.join(char.properties)}), Value: {value}"
+                        )
+
                 else:
                     value = None
-                log.info(
-                    "\t[Characteristic] {0}: (Handle: {1}) ({2}) | Name: {3}, Value: {4} ".format(
-                        char.uuid,
-                        char.handle,
-                        ",".join(char.properties),
-                        char.description,
-                        value,
+                    log.info(
+                        f"\t[Characteristic] {char} ({','.join(char.properties)}), Value: {value}"
                     )
-                )
+
                 for descriptor in char.descriptors:
-                    value = await client.read_gatt_descriptor(descriptor.handle)
-                    log.info(
-                        "\t\t[Descriptor] {0}: (Handle: {1}) | Value: {2} ".format(
-                            descriptor.uuid, descriptor.handle, bytes(value)
+                    try:
+                        value = bytes(
+                            await client.read_gatt_descriptor(descriptor.handle)
                         )
-                    )
+                        log.info(f"\t\t[Descriptor] {descriptor}) | Value: {value}")
+                    except Exception as e:
+                        value = str(e).encode()
+                        log.error(f"\t\t[Descriptor] {descriptor}) | Value: {value}")
```

Comment:
Shouldn't we be able to use `e` directly in the format string?

    log.error(f"\t\t[Descriptor] {descriptor}) | Exception: {e}")

Reply:
Interesting. I wonder why I I thought this was a good idea? I think I just wanted it to be a bytes object like values in general. There is no need for that though.